### PR TITLE
Added ability to output all UMIs to a single tag as well as individua…

### PR DESCRIPTION
**Pull request by tfenne**
…l tags in ExtractUmisFromBam.

### CODE REVIEW

**Comment by tfenne** [comment]
@nh13 There are really two changes in one here.  The functional change is to allow the tool to output a single tag with concatenated UMIs (e.g. `RX`) in addition to per-UMI tags (e.g. `ZA` and `ZB` or similar).  It's only really useful when there are multiple UMIs but it can be used whether or not there are multiple.

In addition it was driving me slightly crazy that the tool was called `ExtractUmisFromBam` where `umi` == "unique molecular _index_" but that a lot of the documentation and variables used `molecularBarcode`.  I changed everything over to use _index_ instead, and deprecated the prior command line parameter.

File: src/main/scala/com/fulcrumgenomics/umi/ExtractUmisFromBam.scala [hunk 1]
```diff
@@ -204,26 +225,26 @@ object ExtractUmisFromBam {
...
-    // get the molecular barcode segments
-    val molecularBarcodeBases = readStructureBases.collect { case SubRead(b: String, _, m: MolecularBarcode) => b }
-    // set the barcode tags
-    molecularBarcodeTags match {
-      case Seq(tag) => record.setAttribute(tag, molecularBarcodeBases.mkString(UmiDelimiter))
+    // get the molecular index segments
+    val molecularIndexBases = readStructureBases.collect { case SubRead(b: String, _, m: MolecularBarcode) => b }
+    // set the index tags
+    molecularIndexTags match {
+      case Seq(tag) => record.setAttribute(tag, molecularIndexBases.mkString(UmiDelimiter))
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/umi/ExtractUmisFromBam.scala]: I think when the `molecularBarcodesTag` option is removed, we'll want to not have the behavior of having the values concatenated if the  `molecularBarcodesTag` has only one tag.  Could you add a note to that effect here?

File: src/test/scala/com/fulcrumgenomics/umi/ExtractUmisFromBamTest.scala [hunk 2]
```diff
@@ -266,7 +266,41 @@ class ExtractUmisFromBamTest extends UnitSpec with OptionValues {
...
+    an[ValidationException] shouldBe thrownBy {
+      new ExtractUmisFromBam(input=in, output=out, molecularIndexTags=Seq("RX", "RZ"), singleTag=Some("RX"), readStructure=Seq("5M5T", "5M5T"))
+    }
+
+    an[ValidationException] shouldBe thrownBy {
+      new ExtractUmisFromBam(input=in, output=out, molecularIndexTags=Seq("RX", "RZ"), singleTag=Some("RXX"), readStructure=Seq("5M5T", "5M5T"))
+    }
+  }
+
+
```
Comment (nh13) [inline src/test/scala/com/fulcrumgenomics/umi/ExtractUmisFromBamTest.scala]: nl

File: src/main/scala/com/fulcrumgenomics/umi/ExtractUmisFromBam.scala (hunk 1 above)
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/umi/ExtractUmisFromBam.scala]: Sure.  Though I think we may still want it....  Without it there's no way to get all the values contatenated in a single tag _without_ also writing per-UMI tags.  The latest version supports three modes:
1) Just write per umi tags by providing the appropriate number of `molecular-index-tag`s
2) Just write a single tag with all values by providing a single `molecular-index-tag`
3) Write per-umi tags _and_ a concatenated tag by providing the appropriate number of `molecular-index-tag`s _and_ `single-tag`.

**Comment by tfenne** [comment]
Agreed on the deprecation making it a bit messy, but I didn't want to just change the option name without warning.  Requested changes are made, will merge once tests pass.

**Pull request merged**